plugins/poll/plugin.rb

```diff
@@ -398,37 +411,72 @@ def voters
   end
 
   on(:post_created) do |post|
-    next if post.is_first_post? || post.custom_fields[DiscoursePoll::POLLS_CUSTOM_FIELD].blank?
-    # signals the front-end we have polls for that post
-    MessageBus.publish("/polls/#{post.topic_id}", post_id: post.id, polls: post.custom_fields[DiscoursePoll::POLLS_CUSTOM_FIELD])
-    # schedule automatic close jobs
     DiscoursePoll::Poll.schedule_jobs(post)
+
+    unless post.is_first_post?
+      polls = ActiveModel::ArraySerializer.new(post.polls, each_serializer: PollSerializer, root: false).as_json
+      MessageBus.publish("/polls/#{post.topic_id}", post_id: post.id, polls: polls)
+    end
   end
 
   on(:merging_users) do |source_user, target_user|
-    DiscoursePoll::VotesUpdater.merge_users!(source_user, target_user)
+    PollVote.where(user: source_user).update_all(user: target_user)
   end
 
-  add_to_serializer(:post, :polls, false) do
-    polls = post_custom_fields[DiscoursePoll::POLLS_CUSTOM_FIELD].dup
+  on(:user_destroyed) do |user|
+    PollVote.where(user: user).delete_all
+  end
 
-    polls.each do |_, poll|
-      poll["options"].each do |option|
-        option.delete("voter_ids")
-      end
+  add_to_class(:topic_view, :polls) do
+    @polls ||= begin
```

Comment:
what we can do here is piggy back on a custom field. So you would store a single custom field on a post saying "hey I have a poll" and then if that hits go to the polls table. This means that we would only run the complex poll query if one of the posts has a poll. This is much more efficient cause we run the "fetch these custom fields query" for multiple plugins. 99% of posts have no polls.

Reply:
I've done that in https://github.com/discourse/discourse/pull/6359/commits/c121b64cee41a33f9c5b5fda1fcf85826968fccf#diff-89c344c997376afa256e678abfef12b0R441